lib/gtl_formatter.rb

```diff
@@ -41,17 +88,102 @@ def self.format_cols(view, row, controller)
         celltext = send(special_cases[view.extras[:filename]], view, row, col)
       elsif special_cases[view.extras[:filename]].kind_of?(Hash) && special_cases[view.extras[:filename]][view.col_order[col_idx]].present?
         celltext, span = send(special_cases[view.extras[:filename]][view.col_order[col_idx]], row[col])
+      elsif COLUMN_WITH_IMAGE.keys.include?(col)
+        # Generate html for the list icon
+        icon, icon2, image = send(COLUMN_WITH_IMAGE[col], record)
+        text = format_col_for_display(view, row, col)
+        item = {:title => text,
+                :image => ActionController::Base.helpers.image_path(image.to_s),
+                :icon  => icon,
+                :icon2 => icon2,
+                :text  => text}.compact
+      elsif COLUMN_WITH_ICON.keys.include?(col)
+        # Generate html for the list icon
+        icon = send(COLUMN_WITH_ICON[col], record)
+        text = format_col_for_display(view, row, col)
+        item = {:title => text,
+                :icon  => icon,
+                :text => text}.compact
+      elsif COLUMN_WITH_TIME.include?(col)
+        celltext = format_time_for_display(row, col)
       else
         celltext = format_col_for_display(view, row, col)
       end
 
-      item = {:text => celltext}
+      item ||= {:text => celltext}
       item[:span] = span if span.present?
       cols.push(item)
     end
+
+    # Append a button if @row_button is set and the button is defined in the related decorator
+    button = record.decorate.try(:gtl_button_cell) if options[:row_button]
+    cols.push(button) if button
+
     cols
   end
 
+  def self.format_icon_column(view, row, clickable)
+    item = listicon_item(view, row['id'])
+    icon, icon2, image = fonticon_or_fileicon(item)
+
+    # Clickable should be false only when it's explicitly set to false
+    {:title => clickable == false ? nil : _('View this item'),
+     :image => ActionController::Base.helpers.image_path(image.to_s),
+     :icon  => icon,
+     :icon2 => icon2}.compact
+  end
+
+  def self.listicon_item(view, id = nil)
+    id = @id if id.nil?
+
+    if @targets_hash
+      @targets_hash[id] # Get the record from the view
+    else
+      klass = view.db.constantize
+      klass.find(id)    # Read the record from the db
+    end
+  end
+
+  def self.fonticon_or_fileicon(item)
+    return nil unless item
+    decorated = item.decorate
+    [
+      decorated.try(:fonticon),
+      decorated.try(:secondary_icon),
+      decorated.try(:fileicon)
+    ]
+  end
+
+  def self.last_compliance_status_image(item)
+    case item.last_compliance_status
```

Comment:
I *think* this is causing the `SELECT  "compliances".* FROM "compliances" WHERE "compliances"."resource_id" = $1 AND "compliances"."resource_type" = $2 ` set of queries, because it's a virtual delegate to `last_compliance.compliant`.

Reply:
I notice we're calling the `COLUMN_WITH_ICON` method with `record`, but we already have `last_compliance_status` in the report.

I think if you just call this with `row` instead of `record`, that particular N+1 will disappear.
(Well, and `row` is a hash, not object)